splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampConnection.java

```diff
@@ -0,0 +1,403 @@
+    private AtomicLong numRequests = new AtomicLong(0);
+    private AtomicLong totalRequestDuration = new AtomicLong(0);
+    private TimestampHostProvider timestampHostProvider;
+    private Bootstrap bootstrap;
+    private EventLoopGroup workerGroup;
+
+
+    public TimestampConnection(int timeoutMillis, TimestampHostProvider timestampHostProvider) {
+        this.timeoutMillis = timeoutMillis;
+        this.timestampHostProvider = timestampHostProvider;
+        clientCallbacks = new ConcurrentHashMap<>();
+
+        workerGroup =  new NioEventLoopGroup(2, new ThreadFactoryBuilder().setNameFormat("TimestampClient-worker-%d").setDaemon(true).build());
+
+        bootstrap = new Bootstrap();
+        bootstrap.group(workerGroup);
+        bootstrap.channel(NioSocketChannel.class);
+        bootstrap.option(ChannelOption.SO_KEEPALIVE, true);
+        bootstrap.option(ChannelOption.SO_REUSEADDR, true);
+        bootstrap.option(ChannelOption.TCP_NODELAY, true);
+        bootstrap.handler(new ChannelInitializer<SocketChannel>() {
+            @Override
+            public void initChannel(SocketChannel ch) throws Exception {
+                ch.pipeline()
+                .addLast("frameDecoder", new LengthFieldBasedFrameDecoder(1048576, 0, 4, 0, 4))
+                .addLast("protobufDecoder", new ProtobufDecoder(TimestampMessage.TimestampResponse.getDefaultInstance()))
+                .addLast("frameEncoder", new LengthFieldPrepender(4))
+                .addLast("protobufEncoder", new ProtobufEncoder())
+                .addLast("handler", TimestampConnection.this);
+            }
+        });
+
+        try {
+            registerJMX();
+        } catch (Exception e) {
+            SpliceLogUtils.error(LOG, "Unable to register TimestampClient with JMX. Timestamps will still be generated but metrics will not be available.");
+        }
+    }
+
+    public void shutdown() {
+        boolean shouldContinue = true;
+        while(shouldContinue){
+            State state=this.state.get();
+            if(state== State.SHUTDOWN) return;
+            shouldContinue=!this.state.compareAndSet(state, State.SHUTDOWN);
+        }
+        LOG.info(String.format("shutting down TimestampClient state=%s", this.state.get()));
+        try {
+            this.state.set(State.SHUTDOWN);
+            workerGroup.shutdownGracefully();
+        } catch (Throwable t) {
+            LOG.error("error shutting down", t);
+        }
+    }
+
+    /**
+     * Returns the port number which the client should use when connecting
+     * to the timestamp server.
+     */
+    protected int getPort() {
+        return timestampHostProvider.getPort();
+    }
+
+    protected void connectIfNeeded() throws TimestampIOException{
+
+        // Even though state is an atomic reference, synchronize on whole block
+        // including code that attempts connection. Otherwise, two threads might
+        // end up trying to connect at the same time.
```

Comment:
Is this comment still relevant?  I don't see the synchronize block.

Reply:
It looks outdated, I moved that directly from TimestampClient, if I get a chance I'll update it. Thanks!